File: HoverFillButton/custombuttonbase.cpp
#include "custombuttonbase.h"

CustomButtonBase::CustomButtonBase(QWidget *parent)
    : QPushButton(parent)
{
    initializeMemberVariable();
}

CustomButtonBase::CustomButtonBase(const QString &text, QWidget *parent)
    : QPushButton(text, parent)
{
    initializeMemberVariable();
}

CustomButtonBase::CustomButtonBase(const QIcon &icon, const QString &text, QWidget *parent)
    : QPushButton(icon, text, parent)
{
    initializeMemberVariable();
}

void CustomButtonBase::addShadow(int right, int down, qreal blurRadius, const QColor &shadowColor)
{
    QGraphicsDropShadowEffect *shadow = new QGraphicsDropShadowEffect(this);
    shadow->setOffset(right, down);    // 阴影的偏移量（右，下）
    shadow->setColor(shadowColor);     // 阴影的颜色
    shadow->setBlurRadius(blurRadius); // 控制阴影的模糊程度（光源距离）
    this->setGraphicsEffect(shadow);
}

void CustomButtonBase::setTextColor(const QColor& normalTextColor, const QColor &hoverTextColor)
{
    hover_text_color = hoverTextColor;
    palette().setColor(QPalette::ButtonText, normalTextColor);
}

QPalette &CustomButtonBase::palette() const
{
    return const_cast<QPalette &>(QPushButton::palette());
}

void CustomButtonBase::refreshRadius(int radius)
{
    this->radius = radius;
    path.clear();
    path.addRoundedRect(rect(), radius, radius); // 重新为绘制路径添加圆角矩形区域
}

void CustomButtonBase::initializeMemberVariable()
{
    path.addRoundedRect(rect(), radius, radius);
}

void CustomButtonBase::resizeEvent(QResizeEvent *event)
{
    QPushButton::resizeEvent(event);
    path.clear();
    path.addRoundedRect(rect(), radius, radius);
}

void CustomButtonBase::enterEvent(QEnterEvent *event)
{
    QPushButton::enterEvent(event);
    is_cursor_inside = true;
}

void CustomButtonBase::leaveEvent(QEvent *event)
{
    QPushButton::leaveEvent(event);
    is_cursor_inside = false;
}

File: Archive scrap code/PySide6/HoverFillButton/CustomButtonBase.py
from PySide6.QtCore import Qt
from PySide6.QtGui import QPainterPath, QPalette
from PySide6.QtWidgets import QPushButton, QGraphicsDropShadowEffect, QWidget


class CustomButtonBase(QPushButton):
    def __init__(self, text: str = None, parent: QWidget = None):
        super().__init__(text=text, parent=parent)
        self.hover_text_color = Qt.GlobalColor.black
        self.path = QPainterPath()
        self.is_cursor_inside = False
        self.radius = 4
        self.path.addRoundedRect(self.rect(), self.radius, self.radius)

    def addShadow(self, right, down, blurRadius, shadowColor=Qt.GlobalColor.black):
        shadow = QGraphicsDropShadowEffect(self)
        shadow.setOffset(right, down)
        shadow.setColor(shadowColor)
        shadow.setBlurRadius(blurRadius)
        self.setGraphicsEffect(shadow)

    def setTextColor(self, normalTextColor, hoverTextColor):
        self.hover_text_color = hoverTextColor
        self.palette().setColor(QPalette.ColorRole.ButtonText, normalTextColor)

    def refreshRadius(self, radius):
        self.radius = radius
        self.path.clear()
        self.path.addRoundedRect(self.rect(), radius, radius)

    def resizeEvent(self, event):
        super().resizeEvent(event)
        self.path.clear()
        self.path.addRoundedRect(self.rect(), self.radius, self.radius)

    def enterEvent(self, event):
        super().enterEvent(event)
        self.is_cursor_inside = True

    def leaveEvent(self, event):
        super().leaveEvent(event)
        self.is_cursor_inside = False
